### PR TITLE
Use HTTPlug

### PR DESCRIPTION
**Pull request by Nyholm**
This PR removes the HTTP adapters and instead we rely on the HTTP client abstraction HTTPlug. This comes with many benefits: 

* we support far more adapters (like Guzzle 6)
* we do use PSR7
* We have less code to test and maintain

### CODE REVIEW

File: lib/Stampie/Mailer.php [hunk 1]
```diff
@@ -2,7 +2,13 @@
 
 namespace Stampie;
 
-use Stampie\Adapter\AdapterInterface;
+use GuzzleHttp\Psr7\MultipartStream;
```
Comment (henrikbjorn) [inline lib/Stampie/Mailer.php]: I don't like relying on Guzzle packages.

Comment (Nyholm) [inline lib/Stampie/Mailer.php]: Neither do I. There is a PR that will create a MultipartStream factory that will abstract away this dependency as well. I'll do my best to make sure it get merged soon. 

Ref: https://github.com/php-http/message-factory/pull/30

**Comment by henrikbjorn** [comment]
@Nyholm not against this change, however we must bump the major version when it gets in. Also the dependency on a Guzzle package limits the usefulness, but when that have been change i am all for merging.

**Comment by Nyholm** [comment]
Thank you. Im working on removing the dependency on Guzzle. I'll let you know.

**Comment by stof** [comment]
Removing adapters can only be done in a major version.
@henrikbjorn is the next release a major version ?

**Comment by henrikbjorn** [comment]
@stof You are of course correct. I would bump the major when we merge this.

**Comment by Nyholm** [comment]
Thank you. I've removed the dependency on Guzzle by introducing a MultipartStreamBuilder that is independent from any Stream implementation.

This PR is now ready for review.

**Comment by henrikbjorn** [comment]
I skimmed the PR and looks fine to me.

@stof do you have any comments/objections etc?

File: lib/Stampie/Mailer.php [hunk 2]
```diff
@@ -24,19 +29,24 @@
...
+     * @var MessageFactory
      */
-    public function __construct(AdapterInterface $adapter, $serverToken)
+    private $messageFactory;
+
+    /**
+     * @param HttpClient $adapter
+     * @param string     $serverToken
+     */
+    public function __construct(HttpClient $adapter = null, $serverToken)
```
Comment (henrikbjorn) [inline lib/Stampie/Mailer.php]: Please don't make this optional. let's force the user to inject what they need, and if they wish they can use the HttpClientDiscovery functionality on their own.

File: composer.json [hunk 3]
```diff
@@ -25,17 +25,20 @@
...
+        "php-http/message-factory"          : "^1.0",
+        "php-http/discovery"                : "^1.0",
+        "php-http/multipart-stream-builder" : "^0.1",
+        "php-http/client-implementation"    : "^1.0"
     },
     "require-dev" : {
-        "kriswallsmith/buzz" : "~0.8",
-        "guzzle/http"        : "~3",
-        "guzzlehttp/guzzle"  : "~5.2"
+        "phpunit/phpunit" : "^4.8",
```
Comment (stof) [inline composer.json]: you should allow both 4.8 and 5.x as 4.8 is not officially supported on PHP 7

File: lib/Stampie/Mailer.php (hunk 2 above)
Comment (stof) [inline lib/Stampie/Mailer.php]: I would keep the argument required here (which could remove the dependency on the client discovery btw)

File: lib/Stampie/Mailer.php [hunk 4]
```diff
@@ -24,19 +29,24 @@
...
-        $this->setAdapter($adapter);
+        $this->setAdapter($adapter ?: HttpClientDiscovery::find());
         $this->setServerToken($serverToken);
     }
 
     /**
      * {@inheritdoc}
      */
-    public function setAdapter(AdapterInterface $adapter)
+    public function setAdapter(HttpClient $adapter)
```
Comment (stof) [inline lib/Stampie/Mailer.php]: the setter should be renamed, as it does not set an adapter anymore

File: lib/Stampie/Mailer.php [hunk 5]
```diff
@@ -50,6 +60,26 @@ public function getAdapter()
     }
 
     /**
+     * @return MessageFactory
+     */
+    private function getMessageFactory()
+    {
+        return $this->messageFactory ?: MessageFactoryDiscovery::find();
```
Comment (henrikbjorn) [inline lib/Stampie/Mailer.php]: Don't know the internals for httplug, but shouldn't it have a method to get the message factory from its adapter? As they are closely related.

File: lib/Stampie/Mailer.php [hunk 6]
```diff
@@ -100,7 +125,9 @@ public function send(MessageInterface $message)
      */
     protected function getHeaders()
     {
-        return array();
+        return array(
+            'Content-Type' => 'application/x-www-form-urlencoded',
+            );
```
Comment (stof) [inline lib/Stampie/Mailer.php]: wrong indentation

File: lib/Stampie/Mailer.php (hunk 5 above)
Comment (henrikbjorn) [inline lib/Stampie/Mailer.php]: Also these are missing from the Interface. However i am not fund of changing the interface too much.

File: lib/Stampie/Mailer.php [hunk 7]
```diff
@@ -173,4 +200,54 @@ protected function buildIdentityString($identities)
...
+        $content = $this->format($message);
+        $headers = $this->getHeaders();
+        $files = $this->getFiles($message);
+
+        if (!empty($files)) {
+            // HTTP query content
+            parse_str($content, $fields);
+            $builder = new MultipartStreamBuilder();
+
+            foreach ($fields as $name => $contents) {
```
Comment (stof) [inline lib/Stampie/Mailer.php]: ``$value`` would be a better name than ``$contents``

File: lib/Stampie/MailerInterface.php [hunk 8]
```diff
@@ -12,12 +12,12 @@
...
+     * @param HttpClient $adapter
      */
-    function setAdapter(AdapterInterface $adapter);
+    function setAdapter(HttpClient $adapter);
 
     /**
-     * @return AdapterInterface
+     * @return HttpClient
      */
     function getAdapter();
```
Comment (stof) [inline lib/Stampie/MailerInterface.php]: @henrikbjorn if we make a new major release, do we really need to have such methods in the interface ? IMO, the HTTP client should be considered an implementation detail rather than being part of the interface

File: tests/Stampie/Tests/MailerTest.php [hunk 9]
```diff
@@ -2,6 +2,9 @@
 
 namespace Stampie\Tests;
 
+
```
Comment (stof) [inline tests/Stampie/Tests/MailerTest.php]: this extra line should be removed IMO

File: lib/Stampie/Mailer.php [hunk 10]
```diff
@@ -173,4 +200,54 @@ protected function buildIdentityString($identities)
...
+                if (!is_array($contents)) {
```
Comment (henrikbjorn) [inline lib/Stampie/Mailer.php]: Removing the `!` and flipping the conditions would make it more read able, also you can remove the `else` part of the conditional by using `continue`.

File: tests/Stampie/Tests/MailerTest.php [hunk 11]
```diff
@@ -77,21 +80,20 @@ public function testUnsuccessfulSendCallsHandle()
         $this->mailer->send($this->getMock('Stampie\MessageInterface'));
     }
 
+    /**
+     * @param bool $isSuccessfull
+     *
+     * @return Response
+     */
     protected function getResponseMock($isSuccessfull)
```
Comment (stof) [inline tests/Stampie/Tests/MailerTest.php]: shouldn't you mock the PSR 7 Response instead ?

File: lib/Stampie/Mailer.php [hunk 12]
```diff
@@ -173,4 +200,54 @@ protected function buildIdentityString($identities)
...
+                    $builder->addResource($key, $contents, $options);
+                }
+            }
+
+            $content = $builder->build();
+            $headers['Content-Type'] = 'multipart/form-data; boundary='.$builder->getBoundary();
+        }
+
+        $request = $this->getMessageFactory()->createRequest('POST', $this->getEndpoint(), $headers, $content);
+        $psr7Response = $this->getAdapter()->sendRequest($request);
```
Comment (henrikbjorn) [inline lib/Stampie/Mailer.php]: why not call it `response`?

File: lib/Stampie/Mailer.php [hunk 13]
```diff
@@ -173,4 +200,54 @@ protected function buildIdentityString($identities)
...
+                    if (!is_numeric($name)) {
+                        $options['filename'] = $name;
+                    }
+                    $contents = fopen($path, 'r');
+                    $builder->addResource($key, $contents, $options);
+                }
+            }
+
+            $content = $builder->build();
+            $headers['Content-Type'] = 'multipart/form-data; boundary='.$builder->getBoundary();
```
Comment (henrikbjorn) [inline lib/Stampie/Mailer.php]: Is this the same for alle Mailers? I think some requires that the attachment be in base64 format in their json request.

File: tests/Stampie/Tests/MailerTest.php [hunk 14]
```diff
@@ -2,6 +2,9 @@
 
 namespace Stampie\Tests;
 
+
+use GuzzleHttp\Psr7\Response;
```
Comment (henrikbjorn) [inline tests/Stampie/Tests/MailerTest.php]: Is it better to use Guzzle instead of https://github.com/php-http/mock-client in tests?

File: lib/Stampie/MailerInterface.php (hunk 8 above)
Comment (henrikbjorn) [inline lib/Stampie/MailerInterface.php]: It's okay for me to remove those. they have been bothering me for a while.

**Comment by stof** [comment]
This PR drops support for PHP 5.3 (by using short array syntax) and maybe also for PHP 5.4 (dunno why you removed it from the Travis matrix). But you have not updated the requirement in the composer.json, which is bad.

File: lib/Stampie/Mailer.php (hunk 2 above)
Comment (Nyholm) [inline lib/Stampie/Mailer.php]: I like the idea of making the argument required. But that does not remove the need for discovery. As the discovery also is used to find StreamFactories and RequestFactories.

File: lib/Stampie/Mailer.php (hunk 13 above)
Comment (stof) [inline lib/Stampie/Mailer.php]: @henrikbjorn if the mailer returns a non-empty array in ``getFiles``, it means they want a multipart submission. Mailers not using multipart submission are handling attachments in the ``format`` method instead, as they are part of the main content (see the Mandrill implementation for instance)

File: lib/Stampie/Mailer.php (hunk 5 above)
Comment (Nyholm) [inline lib/Stampie/Mailer.php]: > Don't know the internals for httplug, but shouldn't it have a method to get the message factory from its adapter? As they are closely related.

The design of classes in httplug follow single responsibility principle. MessageFactories has nothing to do with clients. 

> Also these are missing from the Interface. However i am not fund of changing the interface too much.

This is a private method, there is no need of changing the interface.

File: lib/Stampie/MailerInterface.php (hunk 8 above)
Comment (stof) [inline lib/Stampie/MailerInterface.php]: I would also remove the setters for the server token btw, making things more flexible for each implementation about how authentication is handled.

File: lib/Stampie/Mailer.php (hunk 2 above)
Comment (stof) [inline lib/Stampie/Mailer.php]: I said it removed the need for the *client* discovery

**Comment by Nyholm** [comment]
Yes, it dropps support for 5.3 and 5.4. (But keeps support for PHP5.5 which is EOL). I will update composer.json. Thank you

File: lib/Stampie/Mailer.php (hunk 12 above)
Comment (Nyholm) [inline lib/Stampie/Mailer.php]: I wanted to be clear what kind of response it was. Just one line below is `return new Response`. But in the latter case we are dealing with a StampieResponse.

File: lib/Stampie/MailerInterface.php (hunk 8 above)
Comment (Nyholm) [inline lib/Stampie/MailerInterface.php]: I removed the `getAdapter`, but if you do not mind I suggest removing `setServerToken` would be a different PR.

File: tests/Stampie/Tests/MailerTest.php (hunk 14 above)
Comment (Nyholm) [inline tests/Stampie/Tests/MailerTest.php]: I sometimes use Guzzle/PSR7 just as value objects. But I could rewrite the code to make sure we do not use either.

**Comment by Nyholm** [comment]
Thank you for the review. I've updated or replied on all your comments.

**Comment by Nyholm** [comment]
There, I got some error on hhvm and php5.6 with phpunit 5.4. Build is all green now.

**Comment by Nyholm** [comment]
Is there anything else I can do to help getting this PR merged?

File: composer.json [hunk 15]
```diff
@@ -25,17 +25,20 @@
...
+        "php-http/multipart-stream-builder" : "^0.1",
+        "php-http/client-implementation"    : "^1.0"
     },
     "require-dev" : {
-        "kriswallsmith/buzz" : "~0.8",
-        "guzzle/http"        : "~3",
-        "guzzlehttp/guzzle"  : "~5.2"
+        "phpunit/phpunit" : "^4.8 || 5.0 - 5.3",
+        "php-http/message" : "^1.0",
+        "php-http/guzzle6-adapter" : "^1.0"
```
Comment (henrikbjorn) [inline composer.json]: Is this still required for dev? Can't see it being used in tests anymore.

Comment (Nyholm) [inline composer.json]: Yes, or we need a PSR-7 implementation because we use create Requests and Responses (even in tests) 

We do also need a client-implementation. I will replace that line for 
```json

        "guzzlehttp/psr7" : "^1.3",
        "php-http/mock-client" : "^0.3"
```
to be more clear about this.

**Comment by henrikbjorn** [comment]
@Nyholm Thanks for your work on this.

**Comment by henrikbjorn** [comment]
Have tagget 1.0.0-alpha1.

**Comment by Nyholm** [comment]
Awesome. Thank you for merging
